### PR TITLE
ref(save): Add statement timeout for project counter

### PR DESCRIPTION
**Pull request by beezz**
This PR introduces a configurable statement timeout for the project counter pgsql function call. The timeout should prevent a single project from saturating all save event workers in case of contention on the project counter row due to the row-level locking.

This happens when a project is creating too many groups at the same time, usually due to bad grouping config, unique id/timestamp in event message without stack trace etc.

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -2143,6 +2143,12 @@ def create_partitioned_queues(name):
 
 SENTRY_REPROCESSING_SYNC_REDIS_CLUSTER = "default"
 
+# Timeout for the project counter statement execution.
+# In case of contention on the project counter, prevent workers saturation with
+# save_event tasks from single project.
+# Set to `None` to disable.
+SENTRY_PROJECT_COUNTER_STATEMENT_TIMEOUT = 3000
```
Comment (wedamija) [inline src/sentry/conf/server.py:2150]: Wondering if 3 seconds is actually too long here. I feel like if we're blocking workers, a bunch of things blocking for 3 seconds might still be bad.

**Review by wedamija — APPROVED** [review]
I wasn't totally sure if this would work so I tested it out locally:

psql:
```
BEGIN;
lock table sentry_projectcounter;
```

Verified locked in another psql:
`select * from sentry_projectcounter;` - never responds

Then ran:
```
In [4]: with transaction.atomic():
   ...:     cur = connections['default'].cursor()
   ...:     cur.execute("select sentry_increment_project_counter(1, 1)")
   ...:     cur.close()
```
This hangs forever

```
In [2]: with transaction.atomic():
   ...:     cur = connections['default'].cursor()
   ...:     cur.execute("set local statement_timeout = 3000")
   ...:     cur.execute("select sentry_increment_project_counter(1, 1)")
   ...:     cur.close()
```

This times out after 3s with
```
QueryCanceledError('canceling statement due to statement timeout\nCONTEXT:  SQL statement "update sentry_projectcounter set value = value + delta\n                where project_id = project\n                returning value"\nPL/pgSQL function sentry_increment_project_counter(bigint,integer) line 6 at SQL statement\n',)
SQL: select sentry_increment_project_counter(1, 1)
```

**Review by wedamija — COMMENTED** [review]
What's the impact of erroring during group creation btw? Do we retry later? Do all the retries end up blocking the queue?

File: src/sentry/conf/server.py (hunk 1 above)
Comment (beezz) [reply src/sentry/conf/server.py]: Lowering this down to 1s. We have metrics for this that show that in normal operations call to project counter increment takes on avg 2ms.

**Comment by beezz** [comment]
> What's the impact of erroring during group creation btw? Do we retry later? Do all the retries end up blocking the queue?

The intention is not to retry but shed the load by dropping the event and refund.

File: src/sentry/event_manager.py [hunk 2]
```diff
@@ -936,6 +936,10 @@ def _save_aggregate2(event, flat_hashes, hierarchical_hashes, release, **kwargs)
     existing_group_id = _find_group_id(all_hashes)
 
     if existing_group_id is None:
+
+        if project.id in (options.get("store.load-shed-group-creation-projects") or ()):
+            raise HashDiscarded("Load shedding group creation")
```
Comment (tonyo) [inline src/sentry/event_manager.py:941]: @untitaker Is it applicable here, or had not been added intentionally?

**Review by untitaker — APPROVED** [review]

File: src/sentry/models/counter.py [hunk 3]
```diff
@@ -22,22 +23,28 @@ def increment(cls, project, delta=1):
...
-        cur.execute(
-            """
-            select sentry_increment_project_counter(%s, %s)
-        """,
-            [project.id, delta],
-        )
-        return cur.fetchone()[0]
-    finally:
-        cur.close()
+    # To prevent the statement_timeout leaking into the session we need to use
```
Comment (tonyo) [inline src/sentry/models/counter.py:31]: Please let's add a visible warning that this is not a proper fix and should be removed when we figure out a better way to get the next project id.

**Review by tonyo — APPROVED** [review]

File: src/sentry/models/counter.py (hunk 3 above)
Comment (beezz) [reply src/sentry/models/counter.py:31]: Added warning next to the `statement_timeout` line.

File: src/sentry/event_manager.py (hunk 2 above)
Comment (untitaker) [reply src/sentry/event_manager.py:941]: this is fine

**Pull request merged**